test/toJSON.test.js

```diff
@@ -172,3 +172,1020 @@ test('throw an error on non nullable null object', (t) => {
   })
   t.throws(() => { stringify(null) })
 })
+
+test('recursive toJSON call', (t) => {
+  t.plan(1)
+
+  const schema = {
+    type: 'object',
+    properties: {
+      firstName: {
+        type: 'string'
+      },
+      lastName: {
+        type: ['string', 'null']
+      },
+      age: {
+        type: 'integer',
+        minimum: 0
+      }
+    }
+  }
+
+  const stringify = build(schema)
+
+  const data = {
+    props: {
+      firstName: { toJSON () { return 'Matteo' } },
+      lastName: { toJSON () { return 'Collina' } },
+      age: { toJSON () { return 32 } }
+    },
+    toJSON () {
+      return {
+        firstName: this.props.firstName,
+        lastName: this.props.lastName,
+        age: this.props.age
+      }
+    }
+  }
+
+  t.equal(stringify(data), JSON.stringify(data))
+})
+
+test('toJSON - ref - properties', (t) => {
+  t.plan(2)
+
+  const schema = {
+    title: 'object with $ref',
+    definitions: {
+      def: {
+        type: 'object',
+        properties: {
+          str: {
+            type: 'string'
+          }
+        }
+      }
+    },
+    type: 'object',
+    properties: {
+      obj: {
+        $ref: '#/definitions/def'
+      }
+    }
+  }
+
+  const object = {
+    props: {
+      obj: {
+        toJSON () { return { str: 'test' } }
+      }
+    },
+    toJSON () {
+      return { obj: this.props.obj }
+    }
+  }
+
+  const stringify = build(schema)
+  const output = stringify(object)
+
+  JSON.parse(output)
+  t.pass()
```

Comment:
I would move this line into the `toJSON` method, or it seems not informative here